.github/workflows/main.yml

```diff
@@ -246,19 +233,25 @@ jobs:
       run: rustup update nightly && rustup default nightly
     - run: ./ci/build-std-detect.sh
 
-  success:
+  conclusion:
     needs:
       - docs
       - verify
       - env_override
       - test
       - build-std-detect
     runs-on: ubuntu-latest
-    # GitHub branch protection is exceedingly silly and treats "jobs skipped because a dependency
-    # failed" as success. So we have to do some contortions to ensure the job fails if any of its
-    # dependencies fails.
-    if: always() # make sure this is never "skipped"
+    # We need to ensure this job does *not* get skipped if its dependencies fail,
+    # because a skipped job is considered a success by GitHub. So we have to
+    # overwrite `if:`. We use `!cancelled()` to ensure the job does still not get run
+    # when the workflow is canceled manually.
+    #
+    # ALL THE PREVIOUS JOBS NEED TO BE ADDED TO THE `needs` SECTION OF THIS JOB!
+    if: ${{ !cancelled() }} # make sure this is never "skipped"
     steps:
-      # Manually check the status of all dependencies. `if: failure()` does not work.
-      - name: check if any dependency failed
-        run: jq --exit-status 'all(.result == "success")' <<< '${{ toJson(needs) }}'
+      - name: Conclusion
+        run: |
+          # Print the dependent jobs to see them in the CI log
+          jq -C <<< '${{ toJson(needs) }}'
+          # Check if all jobs that we depend on (in the needs array) were successful.
+          jq --exit-status 'all(.result == "success")' <<< '${{ toJson(needs) }}'
```

Comment:
copy pasted the same check we added in other repos
https://github.com/rust-lang/rust-clippy/blob/8bc1a9e49a22b0e43c942102b13ded8704851228/.github/workflows/clippy_pr.yml#L62